protocol/communities/roles_authorization.go

```diff
@@ -0,0 +1,97 @@
+	protobuf.CommunityMember_ROLE_NONE:             []protobuf.CommunityEvent_EventType{},
+	protobuf.CommunityMember_ROLE_OWNER:            ownerAuthorizedEventTypes,
+	protobuf.CommunityMember_ROLE_MANAGE_USERS:     []protobuf.CommunityEvent_EventType{},
+	protobuf.CommunityMember_ROLE_MODERATE_CONTENT: []protobuf.CommunityEvent_EventType{},
+	protobuf.CommunityMember_ROLE_ADMIN:            adminAuthorizedEventTypes,
+	protobuf.CommunityMember_ROLE_TOKEN_MASTER:     tokenMasterAuthorizedEventTypes,
+}
+
+var adminAuthorizedPermissionTypes = []protobuf.CommunityTokenPermission_Type{
+	protobuf.CommunityTokenPermission_BECOME_MEMBER,
+	protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL,
+	protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL,
+}
+
+var tokenMasterAuthorizedPermissionTypes = append(adminAuthorizedPermissionTypes, []protobuf.CommunityTokenPermission_Type{
+	protobuf.CommunityTokenPermission_BECOME_ADMIN,
+}...)
+
+var ownerAuthorizedPermissionTypes = append(tokenMasterAuthorizedPermissionTypes, []protobuf.CommunityTokenPermission_Type{
+	protobuf.CommunityTokenPermission_BECOME_TOKEN_MASTER,
+}...)
+
+var rolesToAuthorizedPermissionTypes = map[protobuf.CommunityMember_Roles][]protobuf.CommunityTokenPermission_Type{
+	protobuf.CommunityMember_ROLE_NONE:             []protobuf.CommunityTokenPermission_Type{},
+	protobuf.CommunityMember_ROLE_OWNER:            ownerAuthorizedPermissionTypes,
+	protobuf.CommunityMember_ROLE_MANAGE_USERS:     []protobuf.CommunityTokenPermission_Type{},
+	protobuf.CommunityMember_ROLE_MODERATE_CONTENT: []protobuf.CommunityTokenPermission_Type{},
+	protobuf.CommunityMember_ROLE_ADMIN:            adminAuthorizedPermissionTypes,
+	protobuf.CommunityMember_ROLE_TOKEN_MASTER:     tokenMasterAuthorizedPermissionTypes,
+}
+
+func canRolesPerformEvent(roles []protobuf.CommunityMember_Roles, eventType protobuf.CommunityEvent_EventType) bool {
+	for _, role := range roles {
+		authorizedEventTypes := rolesToAuthorizedEventTypes[role]
+		for _, authorizedEventType := range authorizedEventTypes {
+			if authorizedEventType == eventType {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func canRolesModifyPermission(roles []protobuf.CommunityMember_Roles, permissionType protobuf.CommunityTokenPermission_Type) bool {
+	for _, role := range roles {
+		authorizedPermissionTypes := rolesToAuthorizedPermissionTypes[role]
+		for _, authorizedPermissionType := range authorizedPermissionTypes {
+			if authorizedPermissionType == permissionType {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func RolesAuthorizedToPerformEvent(roles []protobuf.CommunityMember_Roles, event *CommunityEvent) bool {
+	if !canRolesPerformEvent(roles, event.Type) {
+		return false
+	}
+
+	if event.Type == protobuf.CommunityEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_CHANGE ||
+		event.Type == protobuf.CommunityEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_DELETE {
+		return canRolesModifyPermission(roles, event.TokenPermission.Type)
```

Comment:
Here we need to check additionally, that Admin did not try to create CommunityTokenPermission_BECOME_TOKEN_MASTER or CommunityTokenPermission_BECOME_TOKEN_ADMIN

Is it TODO?

Reply:
It is not TODO, this method covers exactly this.

https://github.com/status-im/status-go/blob/2a8438e64c0e2bcafb64ee4b49a24e9f89252d18/protocol/communities/roles_authorization.go#L39-L43